src/core/server/preboot/types.ts

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { PluginName } from '..';
+
+/** @internal */
+export interface InternalPrebootServicePreboot {
+  /**
+   * Indicates whether Kibana is currently on hold and cannot proceed to `setup` yet.
+   */
+  readonly isSetupOnHold: () => boolean;
+
+  /**
+   * Registers a `Promise` as a precondition before Kibana can proceed to `setup`. This method can be invoked multiple
+   * times and from multiple `preboot` plugins. Kibana will proceed to `setup` only when all registered `Promise` are
+   * resolved, or it will shut down if any of them is rejected.
+   * @param pluginName Name of the plugin that needs to hold `setup`.
+   * @param reason A string that explains the reason why this promise should hold `setup`. It's supposed to be a human
+   * readable string that will be recorded in the logs or standard output.
```

Comment:
It's a good question. I'm leaning towards it being not i18n-zed, it's supposed to be a service message for the Kibana operator. We tend to keep service messages in English so that it'd be easier to google them if something goes wrong.